### PR TITLE
Created util/instagram.py, modified search/Instagram.py and some smaller changes

### PR DESCRIPTION
**Pull request by rishi23root**
### Updates in this Pull Request
 - update instagram.py
 - create util/instagram.py
 - DRY alert_results function in initial.py
 - use context manager in function - thread and remove unused code in initial.py
 - make the Maryam file run only when calling from the command line
 - solve the problem of the scope of v in Maryam file (hidden yet due to property of closures in python)
 - remove the unused variable from the core.py, initial.py and util/google.py
 - add __init__.py files to create packages from folders
 - No value for argument 'line' in unbound method call from inline 48 by changing the name of the function variable x - X in Maryam file

### CODE REVIEW

File: core/util/instagram.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
+"""
+OWASP Maryam!
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
```
Comment (rachejazz) [inline core/util/instagram.py:6]: I feel instagram is already a module listed in smn's to do list.
Please get an Issue approved by @saeeddhqan before adding on modules.
Helps in saving conflicts
cc: @prince-7

File: core/util/instagram.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+
+import requests
+
+class main:
+    def __init__(self, q, limit=50,session_id=""):
```
Comment (saeeddhqan) [inline core/util/instagram.py]: space error, quote error

File: core/util/instagram.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+            session_id: Your Instagrta session id
+        """
+        
+        self.framework = main.framework
+        
+        self.q = q
+        self.limit = int(limit/50) or 1
+        self.session_id = session_id
+        self.base_url = 'https://www.instagram.com/'
+        self.headers = {
```
Comment (saeeddhqan) [inline core/util/instagram.py]: space error

File: core/util/instagram.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+    def __init__(self, q, limit=50,session_id=""):
+        """ instagram.com search
+            
+            q 		  : The query for search
+            limit	  : The number of details min 50 if exist
+            session_id: Your Instagrta session id
+        """
+        
+        self.framework = main.framework
+
```
Comment (saeeddhqan) [inline core/util/instagram.py]: remove it

File: core/util/instagram.py [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+            
+            q 		  : The query for search
+            limit	  : The number of details min 50 if exist
+            session_id: Your Instagrta session id
+        """
+        
+        self.framework = main.framework
+        
+        self.q = q
+        self.limit = int(limit/50) or 1
```
Comment (saeeddhqan) [inline core/util/instagram.py]: space error

File: core/util/instagram.py [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+        self.session_id = session_id
+        self.base_url = 'https://www.instagram.com/'
+        self.headers = {
+                	    'User-Agent': self.framework.rand_uagent().get,
+                        'Content-Type': 'application/x-www-form-urlencoded',
+                        }
+        if self.session_id:
+            self.headers['Cookie'] = f'sessionid={self.session_id};'
```
Comment (saeeddhqan) [inline core/util/instagram.py]: quote error

File: core/util/instagram.py [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+        # output_data
+        self._userdata = {}
+        self._followers = []
+        self._following = []
+        self._post = []
+
+        self.session = requests.Session()
+
+    def run_crawl(self):
+        self.framework.output(f"[INSTAGRAM] Extracting Data From API")
```
Comment (saeeddhqan) [inline core/util/instagram.py]: use verbose

File: core/util/instagram.py [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+        # 1.1 account details find user on instagram 
+        req = self.get_user_account_info()
+        if req.status_code == 404 or self.data == {} :
```
Comment (saeeddhqan) [inline core/util/instagram.py]: space error

File: core/util/instagram.py [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+            # user not found - show similar account names
+            self.framework.error(f"Account not Found.")
```
Comment (saeeddhqan) [inline core/util/instagram.py]: there's no need to use f""

File: core/util/instagram.py [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+            similar_users = self.get_similar_users()
+            if similar_users:
+                self.framework.heading(f"showing similar users",0)
```
Comment (saeeddhqan) [inline core/util/instagram.py]: quote error. space error

File: core/util/instagram.py [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+
+            for index,user in enumerate(similar_users,1):
+                self.framework.alert(f"{index}. user : {user['username']} --> {self.base_url+user['username']}")
```
Comment (saeeddhqan) [inline core/util/instagram.py]: space error

File: core/util/instagram.py [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+        # 1.1 account details find user on instagram 
+        req = self.get_user_account_info()
+        if req.status_code == 404 or self.data == {} :
+            # user not found - show similar account names
+            self.framework.error(f"Account not Found.")
+            similar_users = self.get_similar_users()
+            if similar_users:
+                self.framework.heading(f"showing similar users",0)
+
+            for index,user in enumerate(similar_users,1):
```
Comment (saeeddhqan) [inline core/util/instagram.py]: space error

File: core/util/instagram.py [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+                self.framework.alert(f"{index}. user : {user['username']} --> {self.base_url+user['username']}")
+            return
+
+        if self.is_private or req.status_code != 200 :
```
Comment (saeeddhqan) [inline core/util/instagram.py]: space error

File: core/util/instagram.py [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+            # return if account is private or request fails
+            if self.is_private : 
+                self.framework.error(f"Account is Private.")
```
Comment (saeeddhqan) [inline core/util/instagram.py]: quote error

File: core/util/instagram.py [hunk 15]
```diff
@@ -0,0 +1,250 @@
...
+                if self.session_id == "":
```
Comment (saeeddhqan) [inline core/util/instagram.py]: ''

File: modules/search/instagram.py [hunk 16]
```diff
@@ -19,60 +19,87 @@
...
 	'version': '0.2',
 	'description': 'Search your query in the Instagram and show the results.',
-	'sources': ('google', 'carrot2', 'bing', 'yippy', 'yahoo', 'millionshort', 'qwant', 'duckduckgo'),
+	'sources': ('google', 'carrot2', 'bing', 'yippy', 'yahoo', 'millionshort', 'qwant', 'duckduckgo','instagram'),
 	'options': (
 		('query', None, True, 'Query string', '-q', 'store', str),
 		('limit', 1, False, 'Search limit(number of pages, default=1)', '-l', 'store', int),
 		('count', 50, False, 'Number of links per page(min=10, max=100, default=50)', '-c', 'store', int),
+		('engine', 'google,instagram', False, 'Engine names for search(default=google, instagram)', '-e', 'store', str),
+		('session_id', None, False, 'Session_id for more details results (default="")', '-s', 'store', str),
```
Comment (saeeddhqan) [inline modules/search/instagram.py]: What is session id?

File: modules/search/instagram.py [hunk 17]
```diff
@@ -19,60 +19,87 @@
...
-		attr = engine(q, limit)
-	else:
-		attr = engine(q)
 
-	attr.run_crawl()
-	LINKS += attr.links
-	PAGES += attr.pages
+	# for instagram
+	if eng == 'instagram':
+		attr = self.instagram(query,limit=limit,session_id=session_id)
```
Comment (saeeddhqan) [inline modules/search/instagram.py]: space error

**Review by saeeddhqan — COMMENTED** [review]
It's huge. I need help! You should have told me that what you want to do.
@k4u5h1k 
@prince-7 
@rachejazz

File: core/util/instagram.py (hunk 1 above)
Comment (saeeddhqan) [reply core/util/instagram.py:6]: No problem. He used api

File: modules/search/instagram.py (hunk 16 above)
Comment (rishi23root) [reply modules/search/instagram.py]: for extracting more data of an account, I use session_id of any account use created when login 
using an account's session_id results will be related to that account (this is the way Instagram works)
and using session_id you will never get an error in response

Comment (saeeddhqan) [reply modules/search/instagram.py]: What is session id? This is a cookie?

File: core/util/instagram.py (hunk 3 above)
Comment (rishi23root) [reply core/util/instagram.py]: @saeeddhqan  where exactly?

File: core/initial.py [hunk 18]
```diff
@@ -243,31 +243,32 @@ def do_workspaces(self, params):
...
+			output = output.strip().replace('\n', ' ').replace('\\x', ' ')
+			self.output(f"{prefix*depth}{output}", color)
 
-	def thread(self, function, thread_count, engines, q, q_formats, limit, count, sources):
-		threadpool = concurrent.futures.ThreadPoolExecutor(max_workers=thread_count)
-		futures = (threadpool.submit(
-			function, self, name, q, q_formats, limit, count) for name in engines if name in sources)
-		for _ in concurrent.futures.as_completed(futures):
-			pass
+	def thread(self, *args):
```
Comment (z3y50n) [inline core/initial.py:267]: i am not sure if it's safe to use just `*args`. Maybe `**args` could be better

File: core/util/instagram.py [hunk 19]
```diff
@@ -0,0 +1,248 @@
...
+            # post 
+            self._userdata['post_count'] = self.data['graphql']['user']['edge_owner_to_timeline_media']['count']
+            # profile_pic
+            self._userdata['profile_pic'] = self.data['graphql']['user']['profile_pic_url_hd']
+            # is a private account
+            self.is_private = self.data['graphql']['user']['is_private']
+        return req
+
+    def get_similar_users(self):
+        '''return 20 similar users'''
```
Comment (z3y50n) [inline core/util/instagram.py]: Use double quotes

File: core/util/instagram.py [hunk 20]
```diff
@@ -0,0 +1,248 @@
...
+        url = 'https://www.instagram.com/web/search/topsearch/?query='
+        r = requests.get(url = url+self.q,headers = self.headers)
+        # return first 20 similar names 
+        return [i['user'] for i in r.json()['users']][:20]
+
+    def get_user_posts(self):
+        '''return boolean : if there is more data to extract'''
```
Comment (z3y50n) [inline core/util/instagram.py]: Double quotes

File: core/util/instagram.py [hunk 21]
```diff
@@ -0,0 +1,248 @@
...
+            else:
+                self.followers_next_page  = None
+                return False
+
+        else:
+            self.framework.error('Request For collecting User Followers is unsuccessful.')
+            return False
+    
+    def get_user_following(self):
+        '''return boolean : if there is more data to extract'''
```
Comment (z3y50n) [inline core/util/instagram.py]: Double quotes

File: maryam [hunk 22]
```diff
@@ -41,37 +32,45 @@ def cui(args, format_help):
...
-		x = x('execute')
-		if option in ['-e', '--execute']:
-			x.onecmd(command)
-			exit(0)
-		elif args[0] in ['-s', '--start']:
-			x.onecmd(command)
-	else:
-		x = x('run')
-	while True:
+		X = x('execute')
```
Comment (z3y50n) [inline maryam]: Capital letters are for data structures like trees

File: maryam [hunk 23]
```diff
@@ -41,37 +32,45 @@ def cui(args, format_help):
...
+		if option in ['-e', '--execute','-s', '--start']:
+			# passing command --> ' '.join(args[1:])
+			X.onecmd(' '.join(args[1:]))
+			if option in ['-e', '--execute']:
+				exit(0)
+	else :
```
Comment (z3y50n) [inline maryam]: No space before `:`

**Comment by z3y50n** [comment]
Also don't make so many irrelevant changes in the same PR since it is hard to review and maybe we want to keep only a few of them. Thanks for your efforts!

**Comment by z3y50n** [comment]
Also after a few tries I am getting this error

image

Does this have to do with the API limit?

File: core/initial.py [hunk 24]
```diff
@@ -243,31 +243,32 @@ def do_workspaces(self, params):
 	#           MODULES 			//
 	# ////////////////////////////////
 
-	def alert_results(self, output, prefix=''):
+	def alert_results(self, output, prefix='\t',depth:int = 0,color='N'):
```
Comment (saeeddhqan) [inline core/initial.py]: space error

File: core/initial.py [hunk 25]
```diff
@@ -243,31 +243,32 @@ def do_workspaces(self, params):
...
-				if isinstance(output[key], dict):
-					self.alert_results(output[key], prefix='\t')
-				elif isinstance(output[key], list):
-					self.alert_results(output[key], prefix='\t')
-				else:
-					self.output(f"\t{output[key]}", 'G')
+			for key, value in output.items():
+				if isinstance(value, list) or isinstance(value, dict):
+					self.alert(f"{prefix*depth}{key.upper()}")
+					self.alert_results(value, prefix=prefix, depth=depth+1, color="G")
```
Comment (saeeddhqan) [inline core/initial.py]: quote error

File: core/util/instagram.py (hunk 3 above)
Comment (saeeddhqan) [reply core/util/instagram.py]: self.headers = {HERE

File: core/util/instagram.py (hunk 13 above)
Comment (saeeddhqan) [reply core/util/instagram.py]: space before :

File: core/util/instagram.py [hunk 26]
```diff
@@ -0,0 +1,248 @@
...
+
+        if self.is_private or req.status_code != 200 :
+            # return if account is private or request fails
+            if self.is_private : 
+                self.framework.error('Account is Private.')
+                if self.session_id == '':
+                    return
+
+            else :
+                self.framework.error(f"Request Fails, Url : {req.url}")
```
Comment (saeeddhqan) [inline core/util/instagram.py]: [INSTAGRAM] Req..

File: core/util/instagram.py [hunk 27]
```diff
@@ -0,0 +1,248 @@
...
+                break
+        
+        for _ in range(self.limit):
+            if not self.get_user_following():
+                break
+
+
+    def get_user_account_info(self):
+        """extracting required user info form request"""
+        req = self.session.get(url=self.base_url + f'{self.q}/?__a=1',headers=self.headers)
```
Comment (saeeddhqan) [inline core/util/instagram.py]: space error, format error

File: core/util/instagram.py [hunk 28]
```diff
@@ -0,0 +1,248 @@
...
+                break
+
+
+    def get_user_account_info(self):
+        """extracting required user info form request"""
+        req = self.session.get(url=self.base_url + f'{self.q}/?__a=1',headers=self.headers)    
+        try : 
+            self.data = req.json()
+        except:
+            self.framework.error('Request Fail !! Too many tries')
```
Comment (saeeddhqan) [inline core/util/instagram.py]: before the !! shouldn't be any space. also [INSTAGRAM]

File: core/util/instagram.py [hunk 29]
```diff
@@ -0,0 +1,248 @@
...
+            # profile_pic
+            self._userdata['profile_pic'] = self.data['graphql']['user']['profile_pic_url_hd']
+            # is a private account
+            self.is_private = self.data['graphql']['user']['is_private']
+        return req
+
+    def get_similar_users(self):
+        '''return 20 similar users'''
+        url = 'https://www.instagram.com/web/search/topsearch/?query='
+        r = requests.get(url = url+self.q,headers = self.headers)
```
Comment (saeeddhqan) [inline core/util/instagram.py]: format error

File: core/util/instagram.py [hunk 30]
```diff
@@ -0,0 +1,248 @@
...
+        # return first 20 similar names 
+        return [i['user'] for i in r.json()['users']][:20]
+
+    def get_user_posts(self):
+        '''return boolean : if there is more data to extract'''
+        url = self.base_url + 'graphql/query/?'
+
+        # set payload - id is target-id  and have max limit is 50 at a request
+        payload = '{' + f'"id":"{self._userdata["id"]}"'+ ',"first":50' + '}'
```
Comment (saeeddhqan) [inline core/util/instagram.py]: format error. hard-coding

File: core/util/instagram.py [hunk 31]
```diff
@@ -0,0 +1,248 @@
...
+        if self.posts_next_page != None:
+            payload = '{' + f'"id":"{self._userdata["id"]}"'+ ',"first":50' + f',"after":"{self.posts_next_page}"' +'}'
```
Comment (saeeddhqan) [inline core/util/instagram.py]: format error. hard-coding

File: core/util/instagram.py [hunk 32]
```diff
@@ -0,0 +1,248 @@
...
+        # return first 20 similar names 
+        return [i['user'] for i in r.json()['users']][:20]
+
+    def get_user_posts(self):
+        '''return boolean : if there is more data to extract'''
+        url = self.base_url + 'graphql/query/?'
+
+        # set payload - id is target-id  and have max limit is 50 at a request
+        payload = '{' + f'"id":"{self._userdata["id"]}"'+ ',"first":50' + '}'
+        if self.posts_next_page != None:
```
Comment (saeeddhqan) [inline core/util/instagram.py]: if self.posts_next_page:...

File: core/util/instagram.py [hunk 33]
```diff
@@ -0,0 +1,248 @@
...
+            payload = '{' + f'"id":"{self._userdata["id"]}"'+ ',"first":50' + f',"after":"{self.posts_next_page}"' +'}'
+
+        # encode the payload and add in params
+        params = {'query_hash':'003056d32c2554def87228bc3fd9668a','variables': payload.encode()}
```
Comment (saeeddhqan) [inline core/util/instagram.py]: space error. apply it for all

File: core/util/instagram.py [hunk 34]
```diff
@@ -0,0 +1,248 @@
...
+        # set payload - id is target-id  and have max limit is 50 at a request
+        payload = '{' + f'"id":"{self._userdata["id"]}"'+ ',"first":50' + '}'
+        if self.posts_next_page != None:
+            payload = '{' + f'"id":"{self._userdata["id"]}"'+ ',"first":50' + f',"after":"{self.posts_next_page}"' +'}'
+
+        # encode the payload and add in params
+        params = {'query_hash':'003056d32c2554def87228bc3fd9668a','variables': payload.encode()}    
+        
+        # geting response
+        data = self.session.get(url,params=params,headers=self.headers).json()
```
Comment (saeeddhqan) [inline core/util/instagram.py]: space error

File: core/util/instagram.py [hunk 35]
```diff
@@ -0,0 +1,248 @@
...
+            else:
+                self.posts_next_page  = None
+                return False
+        else:
+            self.framework.error('Request For collecting User POST is unsuccessful.')
+            return False
+
+    def get_user_followers(self):
+        '''return boolean : if there is more data to extract'''
+        url = self.base_url + 'graphql/query/?'
```
Comment (saeeddhqan) [inline core/util/instagram.py]: format error: f"..."

File: core/util/instagram.py [hunk 36]
```diff
@@ -0,0 +1,248 @@
...
+
+        # set payload - id is target-id  and have max limit is 50 at a request
+        payload = '{' + f'"id":"{self._userdata["id"]}"'+ ',"include_reel":"true","fetch_mutual":"true","first":50'.replace('"true"','true') + '}'
```
Comment (saeeddhqan) [inline core/util/instagram.py]: format error. hard-coding

File: core/util/instagram.py [hunk 37]
```diff
@@ -0,0 +1,248 @@
...
+        if self.followers_next_page != None:
+            payload = '{' + f'"id":"{self._userdata["id"]}"'+ ',"include_reel":"true","fetch_mutual":"true","first":50'.replace('"true"','true') + f',"after":"{self.followers_next_page}"' +'}'
```
Comment (saeeddhqan) [inline core/util/instagram.py]: format error. hard-coding

File: core/util/instagram.py [hunk 38]
```diff
@@ -0,0 +1,248 @@
...
+        
+        # encode the payload and add in params        
+        params = {'query_hash':'c76146de99bb02f6415203be841dd25a','variables': payload.encode()}
+        
+        # geting response
+        data = self.session.get(url,params=params,headers=self.headers).json()
```
Comment (saeeddhqan) [inline core/util/instagram.py]: space error

File: core/util/instagram.py [hunk 39]
```diff
@@ -0,0 +1,248 @@
...
+        '''return boolean : if there is more data to extract'''
+        url = self.base_url + 'graphql/query/?'
+
+        # set payload - id is target-id  and have max limit is 50 at a request
+        payload = '{' + f'"id":"{self._userdata["id"]}"'+ ',"include_reel":"true","fetch_mutual":"true","first":50'.replace('"true"','true') + '}'
+        if self.followers_next_page != None:
+            payload = '{' + f'"id":"{self._userdata["id"]}"'+ ',"include_reel":"true","fetch_mutual":"true","first":50'.replace('"true"','true') + f',"after":"{self.followers_next_page}"' +'}'
+        
+        # encode the payload and add in params        
+        params = {'query_hash':'c76146de99bb02f6415203be841dd25a','variables': payload.encode()}
```
Comment (saeeddhqan) [inline core/util/instagram.py]: space error

File: core/util/instagram.py [hunk 40]
```diff
@@ -0,0 +1,248 @@
...
+                self.followers_next_page  = None
+                return False
+
+        else:
+            self.framework.error('Request For collecting User Followers is unsuccessful.')
+            return False
+    
+    def get_user_following(self):
+        '''return boolean : if there is more data to extract'''
+        url = self.base_url + 'graphql/query/?'
```
Comment (saeeddhqan) [inline core/util/instagram.py]: format error

File: core/util/instagram.py [hunk 41]
```diff
@@ -0,0 +1,248 @@
...
+
+        # set payload - id is target-id  and have max limit is 50 at a request
+        payload = '{' + f'"id":"{self._userdata["id"]}"'+ ',"include_reel":"true","fetch_mutual":"true","first":50'.replace('"true"','true') + '}'
```
Comment (saeeddhqan) [inline core/util/instagram.py]: space error. hard-coding

File: core/util/instagram.py [hunk 42]
```diff
@@ -0,0 +1,248 @@
...
+        if self.following_next_page != None:
+            payload = '{' + f'"id":"{self._userdata["id"]}"'+ ',"include_reel":"true","fetch_mutual":"true","first":50'.replace('"true"','true') + f',"after":"{self.following_next_page}"' +'}'
+        
+        # encode the payload and add in params        
+        params = {'query_hash':'3dec7e2c57367ef3da3d987d89f9dbc8','variables': payload.encode()}
```
Comment (saeeddhqan) [inline core/util/instagram.py]: space error

File: core/util/instagram.py [hunk 43]
```diff
@@ -0,0 +1,248 @@
...
+            # have more followers 
+            if data['data']['user']['edge_follow']['page_info']['has_next_page'] :
+                self.following_next_page = data['data']['user']['edge_follow']['page_info']['end_cursor']
+                return True
+            else:
+                self.following_next_page  = None
+                return False
+
+        else:
+            self.framework.error('Request For collecting User Following is unsuccessful.')
```
Comment (saeeddhqan) [inline core/util/instagram.py]: all of the alerts should have [INSTAGRAM]

File: modules/search/instagram.py [hunk 44]
```diff
@@ -19,60 +19,87 @@
...
+		attr.run_crawl()
+		LINKS += attr.links
+		PAGES += attr.pages
+		USERDATA = attr.userdata
 
 def module_api(self):
 	query = self.options['query']
 	limit = self.options['limit']
 	count = self.options['count']
+	session_id = self.options['session_id'] or ""
```
Comment (saeeddhqan) [inline modules/search/instagram.py]: '' not ""

**Comment by saeeddhqan** [comment]
The first stage is over. I fixed all space errors, template errors, hard coding. The next stage is to test the code. In particular, the core changes.

File: modules/search/instagram.py [hunk 45]
```diff
@@ -19,76 +19,138 @@
 
 meta = {
 	'name': 'Instagram Search',
-	'author': 'Aman Singh',
-	'version': '0.2',
+	'author': 'Rishabh Jain, Aman Singh',
```
Comment (rachejazz) [inline modules/search/instagram.py]: May I ask the collaborator and the maintainer to not allow this? 

If i may defend @prince-7 's efforts on authoring this module, another fellow bug fixer can't have his name up with his! 
If so, then all of us should rewrite modules and rename the original authors! It's not how it works, right?
The main author should always be credited for his thought and valuable addition, not any other consequent bug fixer

Comment (saeeddhqan) [reply modules/search/instagram.py]: depends on @prince-7?

Comment (prince-7) [reply modules/search/instagram.py]: The only issue I have is about uniformity because in the earlier modules only the primary author's name is mentioned. Maybe we can have a 'contributors' section somewhere in the project/readme where all the contributors are named so that their efforts are recognized.

Comment (saeeddhqan) [reply modules/search/instagram.py:22]: good. rishabhjainfinal will add the contributor section.

**Comment by rishi23root** [comment]
@z3y50n - https://github.com/saeeddhqan/Maryam/pull/160#issuecomment-816515700

"Does this have to do with the API limit? " -> no but after too many requests Instagram blocks the request
use -s flag for session id, this will never give you any error until session-id is valid.

File: core/initial.py (hunk 18 above)
Comment (rishi23root) [reply core/initial.py:267]: using (**args) can cause errors on calling from other pre-written modules 
due to key-value pairs format👍
because most of the modules I checked use *args format

File: core/util/instagram.py [hunk 46]
```diff
@@ -0,0 +1,255 @@
...
+
+        self.session = requests.Session()
+
+    def run_crawl(self):
+        self.framework.verbose('[INSTAGRAM] Extracting Data From API')
+        # 1.1 account details find user on instagram 
+        req = self.get_user_account_info()
+        if req.status_code == 404 or self.data == {} :
+            # user not found - show similar account names
+            self.framework.error('Account not Found.')
```
Comment (saeeddhqan) [inline core/util/instagram.py]: Why did you change this line again?!

File: core/util/instagram.py [hunk 47]
```diff
@@ -0,0 +1,255 @@
...
+
+        if self.is_private or req.status_code != 200 :
+            # return if account is private or request fails
+            if self.is_private : 
+                self.framework.error('Account is Private.')
+                if self.session_id == '':
+                    return
+
+            else :
+                self.framework.error(f"Request Fails, Url : {req.url}")
```
Comment (saeeddhqan) [inline core/util/instagram.py]: Why? the [INSTAGRAM API] needs to be the first of all of the messages. I corrected it.

File: core/util/instagram.py [hunk 48]
```diff
@@ -0,0 +1,255 @@
...
+                break
+        
+        for _ in range(self.limit):
+            if not self.get_user_following():
+                break
+
+
+    def get_user_account_info(self):
+        """extracting required user info form request"""
+        req = self.session.get(url=self.base_url + f'{self.q}/?__a=1', headers=self.headers)
```
Comment (saeeddhqan) [inline core/util/instagram.py]: Again

File: core/util/instagram.py [hunk 49]
```diff
@@ -0,0 +1,255 @@
...
+            # post 
+            self._userdata['post_count'] = self.data['graphql']['user']['edge_owner_to_timeline_media']['count']
+            # profile_pic
+            self._userdata['profile_pic'] = self.data['graphql']['user']['profile_pic_url_hd']
+            # is a private account
+            self.is_private = self.data['graphql']['user']['is_private']
+        return req
+
+    def get_similar_users(self):
+        '''return 20 similar users'''
```
Comment (saeeddhqan) [inline core/util/instagram.py]: comments should be with "

File: core/util/instagram.py [hunk 50]
```diff
@@ -0,0 +1,255 @@
...
+        r = requests.get(url=url, headers=self.headers)
+        # return first 20 similar names 
+        return [i['user'] for i in r.json()['users']][:20]
+
+    def get_user_posts(self):
+        '''return boolean : if there is more data to extract'''
+        url = f"{self.base_url}graphql/query/?"
+
+        # set payload - id is target-id  and have max limit is 50 at a request
+        payload = {"id": self._userdata["id"], "first": 50}
```
Comment (saeeddhqan) [inline core/util/instagram.py]: quote error

File: core/util/instagram.py [hunk 51]
```diff
@@ -0,0 +1,255 @@
...
+        if self.posts_next_page  :
+            payload["after"] = self.posts_next_page
```
Comment (saeeddhqan) [inline core/util/instagram.py]: quote error

File: core/util/instagram.py [hunk 52]
```diff
@@ -0,0 +1,255 @@
...
+        else:
+            self.framework.error('Request For collecting User POST is unsuccessful.')
+            return False
+
+    def get_user_followers(self):
+        '''return boolean : if there is more data to extract'''
+        url = f"{self.base_url}graphql/query/?"
+
+        # set payload - id is target-id  and have max limit is 50 at a request
+        payload = {"id": self._userdata["id"], "include_reel": True, "fetch_mutual": True, "first": 50}
```
Comment (saeeddhqan) [inline core/util/instagram.py]: again, quote error

**Review by saeeddhqan — COMMENTED** [review]
I don't review until you correct the format, quote, space errors.

**Comment by rishi23root** [comment]
> I don't review until you correct the format, quote, space errors.

So sorry for this 🙇‍♀️
This time I fix all of them (I hope)
last time I just didn't commit from the newly changed file

File: core/util/instagram.py [hunk 53]
```diff
@@ -0,0 +1,255 @@
...
+        if self.following_next_page  :
+            payload['after'] = self.following_next_page
+        
+        payload = json.dumps(payload)
+        
+        # encode the payload and add in params        
+        params = {'query_hash': '3dec7e2c57367ef3da3d987d89f9dbc8', 'variables': payload.encode()}
+        
+        # geting response
+        data = self.session.get(url,params=params,headers=self.headers).json()
```
Comment (saeeddhqan) [inline core/util/instagram.py]: space error

File: core/util/instagram.py [hunk 54]
```diff
@@ -0,0 +1,255 @@
...
+        # return first 20 similar names 
+        return [i['user'] for i in r.json()['users']][:20]
+
+    def get_user_posts(self):
+        """return boolean : if there is more data to extract"""
+        url = f"{self.base_url}graphql/query/?"
+
+        # set payload - id is target-id  and have max limit is 50 at a request
+        payload = {'id': self._userdata['id'], 'first': 50}
+        if self.posts_next_page  :
```
Comment (saeeddhqan) [inline core/util/instagram.py]: space error

File: core/util/instagram.py [hunk 55]
```diff
@@ -0,0 +1,255 @@
...
+            for node in data['data']['user']['edge_owner_to_timeline_media']['edges']:
+                post = f"{self.base_url}p/{node['node']['shortcode']}"
+                self._post.append(post)
+                
+            # have more posts 
+            if data['data']['user']['edge_owner_to_timeline_media']['page_info']['has_next_page'] :
+                self.posts_next_page = data['data']['user']['edge_owner_to_timeline_media']['page_info']['end_cursor']
+                return True
+            else:
+                self.posts_next_page  = None
```
Comment (saeeddhqan) [inline core/util/instagram.py]: space error

File: core/util/instagram.py [hunk 56]
```diff
@@ -0,0 +1,255 @@
...
+            self.framework.error('[INSTAGRAM] Request For collecting User Followers is unsuccessful.')
+            return False
+    
+    def get_user_following(self):
+        """return boolean : if there is more data to extract"""
+        url = f"{self.base_url}graphql/query/?"
+
+        # set payload - id is target-id  and have max limit is 50 at a request
+        payload = {'id': self._userdata['id'], 'include_reel': True, 'fetch_mutual': True, 'first': 50}
+        if self.following_next_page  :
```
Comment (saeeddhqan) [inline core/util/instagram.py]: space error

Comment (saeeddhqan) [inline core/util/instagram.py]: space error

File: core/util/instagram.py [hunk 57]
```diff
@@ -0,0 +1,255 @@
...
+                        'username': node['node']['username'],
+                        'profile_pic_url': node['node']['profile_pic_url']}
+                self._following.append(user)
+
+            # have more followers 
+            if data['data']['user']['edge_follow']['page_info']['has_next_page'] :
+                self.following_next_page = data['data']['user']['edge_follow']['page_info']['end_cursor']
+                return True
+            else:
+                self.following_next_page  = None
```
Comment (saeeddhqan) [inline core/util/instagram.py]: space error

File: modules/search/instagram.py [hunk 58]
```diff
@@ -19,76 +19,138 @@
...
 	'version': '0.2',
 	'description': 'Search your query in the Instagram and show the results.',
-	'sources': ('google', 'carrot2', 'bing', 'yippy', 'yahoo', 'millionshort', 'qwant', 'duckduckgo'),
+	'sources': ('google', 'carrot2', 'bing', 'yippy', 'yahoo', 'millionshort', 'qwant', 'duckduckgo', 'instagram'),
 	'options': (
 		('query', None, True, 'Query string', '-q', 'store', str),
 		('limit', 1, False, 'Search limit(number of pages, default=1)', '-l', 'store', int),
 		('count', 50, False, 'Number of links per page(min=10, max=100, default=50)', '-c', 'store', int),
+		('engine', 'google,instagram', False, 'Engine names for search(default=google, instagram)', '-e', 'store', str),
+		('session_id', None, False, 'Insta Account Session_id for more details results (default="")', '-s', 'store', str),
```
Comment (z3y50n) [inline modules/search/instagram.py:31]: for more detailed results maybe?

File: modules/search/instagram.py [hunk 59]
```diff
@@ -19,76 +19,138 @@
...
 		('thread', 2, False, 'The number of engine that run per round(default=2)', '-t', 'store', int),
```
Comment (z3y50n) [inline modules/search/instagram.py:32]: number of engines

File: core/util/instagram.py (hunk 54 above)
Comment (rishi23root) [reply core/util/instagram.py]: before : ?

File: modules/search/instagram.py (hunk 58 above)
Comment (rishi23root) [reply modules/search/instagram.py:31]: basically to extract followers and following 👍

Comment (rishi23root) [reply modules/search/instagram.py:31]: @z3y50n you will not get any error of "too many tries" on using session_id

File: modules/search/instagram.py (hunk 59 above)
Comment (rishi23root) [reply modules/search/instagram.py:32]: ?

**Comment by saeeddhqan** [comment]
good job.
Thanks!
